crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs

```diff
@@ -0,0 +1,169 @@
+use rustpython_parser::ast::{Expr, ExprKind, Located, StmtKind};
+
+use crate::checkers::ast::Checker;
+use crate::registry::AsRule;
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::types::{
+    Binding, BindingKind, ClassDef, FunctionDef, Range, RefEquality, Scope, ScopeKind,
+};
+use rustc_hash::FxHashMap;
+
+#[violation]
+pub struct FuncDefinedAbove {
+    pub function_name: String,
+    pub function_def_line: usize,
+}
+
+impl Violation for FuncDefinedAbove {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!(
+            "newspaper style: function {} defined in line {} should be moved down.",
+            self.function_name, self.function_def_line
+        )
+    }
+}
+
+/// NEW100
+pub fn check_function_calls(checker: &mut Checker, func: &Expr) {
+    if let Some(diagnostic) = match_callee_expr(checker, func) {
+        if !checker.settings.rules.enabled(diagnostic.kind.rule()) {
+            return;
+        }
+
+        checker.diagnostics.push(diagnostic);
+    }
+}
+
+fn match_callee_expr(checker: &Checker, func: &Expr) -> Option<Diagnostic> {
+    let kind = &checker.ctx.current_scope().kind;
+    //     "parent scopekind: {kind:?}",
+    //     kind = checker.ctx.current_scope_parent()
+    // );
+    if let ScopeKind::Function(FunctionDef {
+        name: caller_name, ..
+    }) = *kind
+    {
+        let Located { node, .. } = func;
+        match node {
+            ExprKind::Attribute {
+                attr: callee_name, ..
+            } => {
+                if let Some(Scope { bindings, .. }, ..) = checker.ctx.current_scope_parent() {
+                    return check_if_func_defined_above(
+                        checker,
+                        func,
+                        bindings,
+                        callee_name,
+                        caller_name,
+                    );
+                }
+            }
```

Comment:
Do you mind documenting the various cases here? I'm having a little trouble following from the code alone (e.g., which path is intended for which combination of definitions).